File: Algorithm/AlgorithmCSharp/AlgorithmCSharp/Q3.cs
using System;

namespace AlgorithmCSharp
{
    /// <summary>
    /// トロミノパズル
    /// </summary>
    class Q3
    {
        private int trominoCount = 0;

        // マスの状態(0: 何もない、-1: 欠けているマス、それ以外: トロミノ。同じ数字のマスが一つのトロミノを表す) [x][y]
        private int[][] masuList;

        public void Start(int n)
        {
            var nn = (int)Math.Pow(2, n);

            this.masuList = new int[nn][];
            for (var i = 0; i < nn; i++)
            {
                this.masuList[i] = new int[nn];
            }

            // ランダムに壊れているマスを作成する
            var ran = new Random();
            var breakX = ran.Next(0, nn - 1);
            var breakY = ran.Next(0, nn - 1);
            this.masuList[breakX][breakY] = -1;

            this.SetTromino(0, nn - 1, 0, nn - 1);
            foreach (var row in this.masuList)
            {
                foreach (var masu in row)
                {
                    Console.Write(masu);
                }

                Console.WriteLine();
            }
        }

        /// <summary>
        /// すべてのマスが埋まるまで、再帰的にトロミノをセットします。
        /// </summary>
        /// <param name="startX">セットするxの範囲の最小値</param>
        /// <param name="endX">セットするxの範囲の最大値</param>
        /// <param name="startY">セットするyの範囲の最小値</param>
        /// <param name="endY">セットするyの範囲の最大値</param>
        private void SetTromino(int startX, int endX, int startY, int endY)
        {
            this.trominoCount++;

            var length = endY - startY + 1;
            var half = length / 2;

            // 中心4マス
            var centerLX = startX + half - 1;
            var centerRX = startX + half;
            var centerTY = startY + half - 1;
            var centerBY = startY + half;

            // すでに埋まっているマス
            int breakX = 0;
            int breakY = 0;
            for (var i = startX; i <= endX; i++)
            {
                for (var j = startY; j <= endY; j++)
                {
                    if (this.masuList[i][j] != 0)
                    {
                        breakX = i;
                        breakY = j;
                        break;
                    }
                }
            }

            // トロミノをセットするマス(3マス)
            if (((centerLX < breakX) || (centerTY < breakY))
                && this.masuList[centerLX][centerTY] == 0)
            {
                this.masuList[centerLX][centerTY] = this.trominoCount;
            }

            if (((centerLX < breakX) || (centerBY > breakY))
                && this.masuList[centerLX][centerBY] == 0)
            {
                this.masuList[centerLX][centerBY] = this.trominoCount;
            }

            if (((centerRX > breakX) || (centerTY < breakY))
                && this.masuList[centerRX][centerTY] == 0)
            {
                this.masuList[centerRX][centerTY] = this.trominoCount;
            }

            if (((centerRX > breakX) || (centerBY > breakY))
                && this.masuList[centerRX][centerBY] == 0)
            {
                this.masuList[centerRX][centerBY] = this.trominoCount;
            }

            if (length == 2) return;

            // 4分割して再帰
            this.SetTromino(startX, startX + half - 1, startY, startY + half - 1);
            this.SetTromino(startX, startX + half - 1, startY + half, endY);
            this.SetTromino(startX + half, endX, startY, startY + half - 1);
            this.SetTromino(startX + half, endX, startY + half, endY);
        }
    }
}

File: Algorithm/AlgorithmFSharp/AlgorithmFSharp/Q3.fs
// トロミノ
module Q3

open System

// マスの状態を表します
type State = 
    | None
    | Break
    | Tromino of int

let makeCounter () =
    let i = ref 0
    (fun () ->
        i := !i + 1
        State.Tromino !i )

// 新しいトロミノの番号を取得します
let getNewTrominoNumber = makeCounter ()

// ランダムに選択した一マスを埋めます
let breakARandomGrid (gridList :State [][]) =
    let rowLength = Array.length gridList
 
    let rand = new Random();
    let x = rand.Next(0, rowLength - 1)
    let y = rand.Next(0, rowLength - 1)

    gridList.[x].[y] <- State.Break

// 指定した範囲における中心座標4点を取得します
let getCenter startX endX startY endY =
    let sideLength = endY - startY + 1
    let half = sideLength / 2

    startX + half - 1, startX + half,  startY + half - 1, startY + half

// 指定された範囲内にあるすでに埋まっているマスの座標を取得します
let getBreakCoord (gridList:State [][]) startX endX startY endY =
    [startX .. endX]
    |> List.collect(fun x -> [startY .. endY] |> List.map(fun y -> x, y))
    |> List.find(fun (x, y) -> not(gridList.[x].[y] = State.None))

// トロミノをセットするべき座標を計算します
let computeSetTrominoCoordList (gridList:State [][]) startX endX startY endY = 
    let centerLX, centerRX, centerBY, centerTY = getCenter startX endX startY endY

    let breakX, breakY = getBreakCoord gridList startX endX startY endY

    seq{
        if breakX > centerLX then
            yield centerLX, centerTY
            yield centerLX, centerBY
    
        if breakX < centerRX then
            yield centerRX, centerTY
            yield centerRX, centerBY

        if breakY > centerBY then
            yield centerLX, centerBY
            yield centerRX, centerBY

        if breakY < centerTY then
            yield centerLX, centerTY
            yield centerRX, centerTY }

// 指定した範囲にトロミノをセットします
let rec setTromino (gridList:State [][]) startX endX startY endY =
    // トロミノをセットする座標の一覧
    let setTrominoCoordList = computeSetTrominoCoordList gridList startX endX startY endY

    // トロミノをセット
    let trominoNumber = getNewTrominoNumber()
    setTrominoCoordList
    |> Seq.iter(fun (x, y) -> gridList.[x].[y] <- trominoNumber)

    let centerLX, centerRX, centerBY, centerTY = getCenter startX endX startY endY

    // 盤面を4分割して再帰
    match endX - startX + 1 with
    | 2 -> ()
    | _ ->
        setTromino gridList startX centerLX startY centerBY
        setTromino gridList startX centerLX centerTY endY
        setTromino gridList centerRX endX startY centerBY
        setTromino gridList centerRX endX centerTY endY

let start (n: int) = 
    let rowLength = int(Math.Pow(2.0, float n))

    // マスの状態(0: 何もない、-1: 欠けているマス、それ以外: トロミノ。同じ数字のマスが一つのトロミノを表す) [x][y]
    let gridList = Array.init rowLength (fun _ -> Array.init rowLength (fun _-> State.None))
    breakARandomGrid gridList

    let lastIndex = rowLength - 1
    setTromino gridList 0 lastIndex 0 lastIndex
    
    let stateToString state =
        let displayText =
            match state with
            | State.None -> "N"
            | State.Break -> "B"
            | State.Tromino n -> string n

        displayText.PadLeft 2

    gridList
        |> Array.iter(fun row -> row |> Array.map stateToString |> printfn "%A")


